### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by jowezarek**
Adds an (up to docstrings) almost complete developers documentation to the library.
Meant to replace the empty documentation that comes with PR #299.
Note that the paths in .github/workflows/documentation.yml are different from the ones in PR #299.
Update: Additionally includes a short guide on how to maintain the documentation.

To locally inspect the documentation:
```
> cd .../psydac/docs
> python3 -m venv .venv
> source .venv/bin/activate
> pip install sphinx
> pip install numpydoc
> pip install pydata_sphinx_theme
# plus eventually additional packages upon failure to do the following:
> make html
# once all necessary packages have been installed and `make html` did its thing, do:
> cd build/html
> open index.html
```

### CODE REVIEW

File: psydac/api/equation.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
 from psydac.fem.basic                import FemField
 from psydac.linalg.solvers           import inverse
 
-__all__ = ('DiscreteEquation',)
+__all__ = ('l2_boundary_projection', 'DiscreteEquation',)
```
Comment (jowezarek) [inline psydac/api/equation.py:17]: @yguclu I will replace all such instances of round parentheses with bracket parentheses tomorrow.

Comment (jowezarek) [reply psydac/api/equation.py]: 750ffe893abcb82fcbc8c358ab4128932c29fc3b

**Review by FrederikSchnack — COMMENTED** [review]
LGTM

**Comment by jowezarek** [comment]
> 

That's true, thank you! c9f11a1a8edbadae1c8e89710f613a1da828308a

File: psydac/api/equation.py (hunk 1 above)
Comment (yguclu) [reply psydac/api/equation.py]: Why do you prefer a list to a tuple?

File: psydac/pyccel/ast/core.py [hunk 2]
```diff
@@ -3413,8 +3413,8 @@ def __init__(
 
         # arguments
 
-        if not iterable(arguments):
-            raise TypeError('arguments must be an iterable')
+        #if not iterable(arguments):
+        #    raise TypeError('arguments must be an iterable')
```
Comment (yguclu) [inline psydac/pyccel/ast/core.py:3417]: What is the reason for this change?

File: psydac/pyccel/ast/numpyext.py [hunk 3]
```diff
@@ -1305,19 +1305,19 @@ class Int64(NumpyInt):
...
+#            FunctionDef('min',[],[],body=[],
+#                decorators={'numpy_wrapper':NumpyMin}),
+#            FunctionDef('max',[],[],body=[],
+#                decorators={'numpy_wrapper':NumpyMax}),
+#            FunctionDef('imag',[],[],body=[],
+#                decorators={'property':'property', 'numpy_wrapper':Imag}),
+#            FunctionDef('real',[],[],body=[],
+#                decorators={'property':'property', 'numpy_wrapper':Real}),
+#            FunctionDef('diagonal',[],[],body=[],
+#                decorators={'numpy_wrapper':Diag})])
```
Comment (yguclu) [inline psydac/pyccel/ast/numpyext.py:1323]: Was this dead code?

File: psydac/api/equation.py (hunk 1 above)
Comment (yguclu) [reply psydac/api/equation.py]: Yes please. `__all__` should be a tuple.

File: docs/index.html
Comment (yguclu) [inline docs/index.html:1]: Do we still need this file? Is it not automatically generated now?

File: docs/source/wip.gif
Comment (yguclu) [inline docs/source/wip.gif:1]: Is this gif used now?

**Review by yguclu — CHANGES_REQUESTED** [review]

File: docs/source/wip.gif
Comment (jowezarek) [reply docs/source/wip.gif:1]: > Is this gif used now?

Yes, as it is included in `docs/source` unlike in the devel branch, were I forgot to add it.

File: docs/index.html
Comment (jowezarek) [reply docs/index.html:1]: No, I will delete it.

Comment (jowezarek) [reply docs/index.html:1]: ac4785fbaa0da0e9a51bfb40d82b18cee91a1643

File: psydac/utilities/vtk.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
                         )
 from pyevtk.hl import _addDataToParallelFile
 
+__all__ = ['writeParallelVTKUnstructuredGrid']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/utilities/vtk.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/utilities/utils.py [hunk 5]
```diff
@@ -4,21 +4,25 @@
 
 import numpy as np
 
+__all__ = ['refine_array_1d', 'unroll_edges', 'split_space', 'split_field', 'animate_field']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/utilities/utils.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/utilities/quadratures.py [hunk 6]
```diff
@@ -9,6 +9,7 @@
 
 import numpy as np
 
+__all__ = ['gauss_legendre', 'gauss_lobatto', 'quadrature']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/utilities/quadratures.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/linalg/topetsc.py [hunk 7]
```diff
@@ -6,6 +6,8 @@
 
 from mpi4py import MPI
 
+__all__ = ['flatten_vec', 'vec_topetsc', 'mat_topetsc']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/linalg/topetsc.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/fem/vector.py [hunk 8]
```diff
@@ -18,6 +18,8 @@
                                  pushforward_2d_hcurl,
                                  pushforward_3d_hcurl)
 
+__all__ = ['VectorFemSpace', 'ProductFemSpace']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/fem/vector.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/fem/tensor.py [hunk 9]
```diff
@@ -40,6 +40,7 @@
                                  eval_fields_3d_weighted,
                                  eval_fields_3d_irregular_weighted)
 
+__all__ = ['TensorFemSpace']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/fem/tensor.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/fem/partitioning.py [hunk 10]
```diff
@@ -9,20 +9,24 @@
 from psydac.core.bsplines  import elements_spans
 from psydac.fem.vector     import ProductFemSpace, VectorFemSpace
 
+__all__ = ['partition_coefficients', 'construct_connectivity', 'get_minus_starts_ends', 'get_plus_starts_ends',
+           'create_cart', 'construct_interface_spaces', 'construct_reduced_interface_spaces']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/fem/partitioning.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/feec/pushforward.py [hunk 11]
```diff
@@ -17,6 +17,8 @@
 # not the jacobian determinant. For this reason, sign descrepancies can
 # happen when comparing against algorithms which use the latter.
 
+__all__ = ['Pushforward']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/feec/pushforward.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/feec/global_projectors.py [hunk 12]
```diff
@@ -14,6 +14,11 @@
 
 from abc import ABCMeta, abstractmethod
 
+__all__ = ['GlobalProjector', 'Projector_H1', 'Projector_Hcurl', 'Projector_Hdiv', 'Projector_L2',
+           'evaluate_dofs_1d_0form', 'evaluate_dofs_1d_1form',
+           'evaluate_dofs_2d_0form', 'evaluate_dofs_2d_1form_hcurl', 'evaluate_dofs_2d_1form_hdiv', 'evaluate_dofs_2d_2form',
+           'evaluate_dofs_3d_0form', 'evaluate_dofs_3d_1form', 'evaluate_dofs_3d_2form', 'evaluate_dofs_3d_3form']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/feec/global_projectors.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/feec/derivatives.py [hunk 13]
```diff
@@ -13,7 +13,7 @@
 from psydac.linalg.basic    import LinearOperator
 from psydac.ddm.cart        import DomainDecomposition, CartDecomposition
 
-__all__ = (
+__all__ = [
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/feec/derivatives.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/grid.py [hunk 14]
```diff
@@ -10,6 +10,10 @@
 from psydac.fem.tensor             import TensorFemSpace
 from psydac.fem.vector             import ProductFemSpace, VectorFemSpace
 
+__all__ = ['get_points_weights', 'create_collocation_basis',
+           'QuadratureGrid', 'BasisValues', 'CollocationBasisValues']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/grid.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/glt.py [hunk 15]
```diff
@@ -32,6 +32,9 @@
 import random
 from mpi4py import MPI
 
+__all__ = ['GltBasicCodeGen', 'DiscreteGltExpr']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/glt.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/fem.py [hunk 16]
```diff
@@ -31,12 +31,11 @@
...
-    'DiscreteBilinearForm',
-    'DiscreteFunctional',
-    'DiscreteLinearForm',
-    'DiscreteSumForm',
-)
+__all__ = ['collect_spaces', 'compute_diag_len', 'get_quad_order', 
+           'construct_test_space_arguments', 'construct_trial_space_arguments', 
+           'construct_quad_grids_arguments', 'reset_arrays', 'do_nothing', 'extract_stencil_mats', 
+           'DiscreteBilinearForm', 'DiscreteFunctional', 'DiscreteLinearForm', 'DiscreteSumForm'
+]
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/fem.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/feec.py [hunk 17]
```diff
@@ -10,7 +10,7 @@
 from psydac.feec.pull_push         import pull_2d_h1, pull_2d_hcurl, pull_2d_hdiv, pull_2d_l2
 from psydac.feec.pull_push         import pull_3d_h1, pull_3d_hcurl, pull_3d_hdiv, pull_3d_l2
 
-__all__ = ('DiscreteDerham',)
+__all__ = ['DiscreteDerham']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/feec.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/expr.py [hunk 18]
```diff
@@ -19,6 +19,7 @@
 from psydac.fem.tensor        import TensorFemSpace
 from psydac.fem.vector        import ProductFemSpace
 
+__all__ = ['DiscreteExpr']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/expr.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/essential_bc.py [hunk 19]
```diff
@@ -7,8 +7,8 @@
 from psydac.linalg.kron    import KroneckerDenseMatrix
 from psydac.linalg.block   import BlockVector, BlockLinearOperator
 
-__all__ = ('apply_essential_bc',)
-
+__all__ = ['apply_essential_bc', 'check_boundary_type', 'apply_essential_bc_kronecker_dense_matrix', 'apply_essential_bc_stencil', 
+           'apply_essential_bc_BlockLinearOperator', 'apply_essential_bc_BlockVector']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/essential_bc.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/equation.py [hunk 20]
```diff
@@ -14,7 +14,7 @@
 from psydac.fem.basic                import FemField
 from psydac.linalg.solvers           import inverse
 
-__all__ = ('DiscreteEquation',)
+__all__ = ['l2_boundary_projection', 'DiscreteEquation']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/equation.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/discretization.py [hunk 21]
```diff
@@ -40,7 +40,7 @@
 from psydac.cad.geometry     import Geometry
 from psydac.mapping.discrete import NurbsMapping
 
-__all__ = ('discretize',)
+__all__ = ['discretize', 'discretize_derham', 'reduce_space_degrees', 'discretize_space', 'discretize_domain']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/discretization.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/basic.py [hunk 22]
```diff
@@ -17,7 +17,7 @@
 from psydac.api.settings        import PSYDAC_BACKENDS, PSYDAC_DEFAULT_FOLDER
 from psydac.api.utilities       import mkdir_p, touch_init_file, random_string, write_code
 
-__all__ = ('BasicCodeGen', 'BasicDiscrete')
+__all__ = ['BasicCodeGen', 'BasicDiscrete']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/basic.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/api/postprocessing.py [hunk 23]
```diff
@@ -63,7 +63,7 @@ def get_grid_lines_2d(domain_h, V_h, *, refine=1):
             - 'eta2' : value of eta2 on the curve
             - 'x'    : x coordinates of N points along the curve
             - 'y'    : y coordinates of N points along the curve
-
+
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/postprocessing.py]: 4bc035e45bafc6b58ab1912794f74f02d5abbcf3

File: psydac/api/postprocessing.py [hunk 24]
```diff
@@ -296,9 +296,9 @@ def add_spaces(self, **femspaces):
         femspaces: dict
             Named femspaces
 
-        Note
-        ----
-        Femspaces are added to ``self._space_info``.
+        Note:
+            * Femspaces are added to ``self._space_info``.
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/api/postprocessing.py]: 4bc035e45bafc6b58ab1912794f74f02d5abbcf3

File: psydac/feec/pull_push.py [hunk 25]
```diff
@@ -2,7 +2,7 @@
 
 from sympde.topology.callable_mapping import BasicCallableMapping
 
-__all__ = (
+__all__ = [
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/feec/pull_push.py]: 4bc035e45bafc6b58ab1912794f74f02d5abbcf3

File: psydac/fem/projectors.py [hunk 26]
```diff
@@ -4,6 +4,7 @@
 from psydac.core.bsplines   import hrefinement_matrix
 from psydac.linalg.stencil  import StencilVectorSpace
 
+__all__ = ['knots_to_insert', 'knot_insertion_projection_operator']
```
[the remark this answers is not in the document]
Comment (jowezarek) [reply psydac/fem/projectors.py]: de89cafe765344393b4700fe021acacbaa16da94

File: psydac/pyccel/ast/core.py (hunk 2 above)
Comment (jowezarek) [reply psydac/pyccel/ast/core.py:3417]: Sphinx imports all of the files (and at least also those imports of each file that are psydac modules) for which documentation is generated. When trying to document several different files, every time I got a similar error message, ending with this particular TypeError. 

When trying to import said files in ipython, this TypeError was not raised :man_shrugging:

File: psydac/pyccel/ast/numpyext.py (hunk 3 above)
Comment (jowezarek) [reply psydac/pyccel/ast/numpyext.py:1323]: I want to say yes. When looking for `NumpyArrayClass` in vs code I find nothing except for the definition itself. At the same time, these lines were (partly) responsible for the import issues described above, as a TypeError in the class __init__ of FunctionDef was raised multiple times.

File: docs/source/maintenance.rst [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+
+For the sake of maintenance, only the /source folder is of interest.
+
+docs/source
+###########
+
+This folder gives structure to the documentation. As this structure is self made, it needs to be maintained as Psydac grows.
+
+   * Every Sphinx documentation must contain an index.rst. Sphinx will generate index.html from this file, the index page of this documentation.
+   * This index.rst file must contain a toctree (.. toctree::) containing all "content" that is not included on the index page itself. As for now, 06.06.2023, this amounts to three sections, namely modules, examples and maintenance. Each of these sections has its own .rst file, modules.rst, examples.rst and maintenance.rst respectively. All .rst files which don't through toctrees have a path to this root toctree will not be part of the documentation.
```
Comment (yguclu) [inline docs/source/maintenance.rst:29]: What do you mean with "don't through toctrees"? It seems to me that we can remove the words "through toctrees".

File: docs/source/maintenance.rst [hunk 28]
```diff
@@ -0,0 +1,164 @@
+Docs Maintenance
+================
+
+Even though the content of this documentation is generated automatically, its structure is handwritten and, unless we use custom sphinx templates in the future, must be maintained.
+
+This section shall serve as a short guide on how to maintaine this documentation.
```
Comment (yguclu) [inline docs/source/maintenance.rst:6]: Typo: maintaine -> maintain

File: docs/source/maintenance.rst [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+Such changes are quickly applied to the documentation, as such changes must be addressed in one file only, see :ref:`here <simpleadd>`. The file in question is the .rst file of the corresponding submodule and it contains all classes and (public) functions of such a submodule.
+
+.. _simpleadd:
+
+Adding or deleting a class or a (public) function to a module
+-------------------------------------------------------------
+
+1) Adding a class
+
+With a coming PR, the class GMRES will be added to the psydac.linalg.solvers submodule.
```
Comment (yguclu) [inline docs/source/maintenance.rst:59]: AFAIK `psydac.linalg` is a **subpackage** of `psydac`, and `psydac.linalg.solvers` is a **module**, i.e. it corresponds to an actual Python file.

File: docs/source/maintenance.rst [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+This modules , prior to the PR, consists of one function, inverse, as well as six classes, ConjugateGradient, PConjugateGradient, ... and LSMR.
```
Comment (yguclu) [inline docs/source/maintenance.rst:60]: Please use inline formatting for any code, including the class names.

The same comment applies to multiple locations in this file.

**Review by yguclu — CHANGES_REQUESTED** [review]
Excellent job @jowezarek! I have a few minor comments

File: docs/source/maintenance.rst (hunk 27 above)
Comment (jowezarek) [reply docs/source/maintenance.rst:29]: I was trying to explain that even though for example docs/source/modules/linalg/solvers.rst is not included in the "root toctree" located in docs/source/index.rst, it is still included in the documentation as there is "a path to the root toctree", i.e., in docs/source/modules/linalg.rst there is a toctree containing .../solvers.rst. And the .../linalg.rst is contained in the docs/source/modules.rst toctree, and .../modules.rst is included in the "root toctree".

As you can tell, I am certainly lacking the right words to explain what I mean :')

Anyways, it's probably clear what is meant even without "through toctrees" and I will remove it.

File: docs/source/maintenance.rst (hunk 29 above)
Comment (jowezarek) [reply docs/source/maintenance.rst:59]: Ah okay, good to know! Will update throughout the documentation.

File: docs/source/maintenance.rst (hunk 27 above)
Comment (yguclu) [reply docs/source/maintenance.rst:29]: So you want to say that the "root toctree" does not have to point to every single module, but it could just point to subpackages of Psydac. For each subpackage one can then provide a "local toctree" which points to the modules to be documented in that subpackage. Correct?

File: psydac/pyccel/ast/core.py (hunk 2 above)
Comment (yguclu) [reply psydac/pyccel/ast/core.py:3417]: All right. Can you please add a dated comment to the code?

File: docs/source/maintenance.rst (hunk 30 above)
Comment (jowezarek) [reply docs/source/maintenance.rst]: 3fa4e0272e920da64a042cbb87deb679ade437c5

I also used inline formatting for instances similar to `psdac.fem` and `block.rst`. Let me know if you are okay with that.

File: docs/source/maintenance.rst (hunk 29 above)
Comment (jowezarek) [reply docs/source/maintenance.rst]: 3fa4e0272e920da64a042cbb87deb679ade437c5

File: docs/source/maintenance.rst (hunk 28 above)
Comment (jowezarek) [reply docs/source/maintenance.rst]: 3fa4e0272e920da64a042cbb87deb679ade437c5

File: docs/source/maintenance.rst (hunk 27 above)
Comment (jowezarek) [reply docs/source/maintenance.rst]: Yes exactly!

Comment (jowezarek) [reply docs/source/maintenance.rst]: 3fa4e0272e920da64a042cbb87deb679ade437c5

File: psydac/pyccel/ast/core.py (hunk 2 above)
Comment (jowezarek) [reply psydac/pyccel/ast/core.py:3418]: dd499a8917e7f04f6a736fb765b69224630f574e

File: docs/source/maintenance.rst [hunk 31]
```diff
@@ -151,14 +155,17 @@ As fem.basic consists of two classes only, FemSpace and FemField, the basic.rst
...
+
+2) Deleting
 
 Undo the above.
 
 .. _nochange:
 
-Changing a class or a function within module
---------------------------------------------
+Changing a class or a (public) function within module
```
Comment (yguclu) [inline docs/source/maintenance.rst:166]: Missing "a": "within a module"

File: docs/source/maintenance.rst [hunk 32]
```diff
@@ -91,39 +93,41 @@ Upon adding GMRES, the docs/source/modules/linalg/solvers.rst file has to be cha
...
+   .. _gmres:						                     # this line is new
    
-   GMRES						# this line is new
-   -----						# this line is new
+   GMRES						                           # this line is new
+   -----						                           # this line is new
    
    .. autoclass:: psydac.linalg.solvers.GMRES		# this line is new
-      :members:						# this line is new
+      :members:						                  # this line is new
```
Comment (yguclu) [inline docs/source/maintenance.rst]: Are the last changes necessary? If so why?

File: docs/source/maintenance.rst (hunk 30 above)
Comment (yguclu) [reply docs/source/maintenance.rst]: That sounds good to me. But how can I see the maintenance page in the generated documentation?

File: docs/source/maintenance.rst (hunk 32 above)
Comment (jowezarek) [reply docs/source/maintenance.rst]: No, I thought the indentation was wrong, but it has actually already been correct.
09d7d3ce32159da9d13294910b7ee016a8b8d579

File: docs/source/maintenance.rst (hunk 31 above)
Comment (jowezarek) [reply docs/source/maintenance.rst]: 09d7d3ce32159da9d13294910b7ee016a8b8d579
